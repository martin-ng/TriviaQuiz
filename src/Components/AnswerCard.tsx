import React, { FC } from 'react'
import '../css/style.css'
import '../css/answer-button.css'

type Props = {
    styling: any,
    answer: string,
    nextQuestion: any,
    userAnswer: any
}

const AnswerCard: FC<Props> = ({ styling, answer, nextQuestion, userAnswer }) => {

    const classes = userAnswer ? `${styling} btn btn-cursor` : `${styling} btn btn-styling btn-cursor`

    return (
        <div >
            <button className={classes} disabled={userAnswer} onClick={() => nextQuestion(answer)}>{answer}</button>
        </div>
    )
}

export default AnswerCard