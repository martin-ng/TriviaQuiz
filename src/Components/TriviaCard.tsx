import React, { FC, useState, useEffect } from 'react'

// Type Checking 
import { QuestionBankState } from '../API'

// Components
import { AnswerCard } from './index'

// Randomize the questions
import { fetchQuestions } from '../API'

// Question Bank
import jsonData from '../questions/tandem_data.json'

import '../css/normal-button.css'

type Props = {
    setGameStatus: any;
}

const MAX_QUESTIONS = 10

const TriviaCard: FC<Props> = ({ setGameStatus }) => {

    const [gameOver, setGameOver] = useState<boolean>(true)
    const [questionBank, setQuestionBank] = useState<QuestionBankState[]>([])
    const [number, setNumber] = useState<number>(0)
    const [score, setScore] = useState<number>(0)
    const [userCorrect, setUserCorrect] = useState<boolean>(false)
    const [userAnswer, setUserAnswer] = useState<string>('')

    useEffect(() => {
        resetState()
    }, [])

    const resetState = () => {
        console.log('fetching questions')
        setQuestionBank(fetchQuestions(jsonData, MAX_QUESTIONS))
        setNumber(0)
        setScore(0)
        setGameOver(false)
    }

    // This function determines the `AnswerCard's` styling depending on correctness.
    const checkColor = (answer: string) => {
        if (userCorrect && answer === userAnswer) return 'correct-answer'
        if (userCorrect || userAnswer === '') return 'choice'
        if (userCorrect && answer !== userAnswer) return 'choice'
        if (!userCorrect && answer === userAnswer) return 'incorrect-answer'
        if (!userCorrect && answer === questionBank[number].correct) return 'correct-answer'
    }

    const checkAnswer = (correct: string, answer: string) => {
        if (answer === correct) {
            setUserCorrect(true)
            setScore(score + 1)
        }
    }

    const isCorrectAnswer = () => {
        return userCorrect ? 'That is Correct!' : 'That is Incorrect.'
    }

    const nextQuestion = (answer: string) => {
        checkAnswer(questionBank[number].correct, answer)
        setUserAnswer(answer)

        setTimeout(() => {
            if (number === MAX_QUESTIONS - 1) {
                setGameOver(true)
            }
            setNumber(number + 1)
            setUserAnswer('')
            setUserCorrect(false)
        }, 1000)
    }

    return (
        <div className='trivia-container'>

            {!gameOver ?
                <div className='question-container'>
                    <p className='question-number'>Question: {number + 1} / {MAX_QUESTIONS}</p>
                    <h3 className='question'>{questionBank[number].question}</h3>
                    {questionBank[number].answers.map((answer: string) => <AnswerCard key={answer} styling={checkColor(answer)} answer={answer} nextQuestion={nextQuestion} userAnswer={userAnswer} />)}
                    {userAnswer ? <p className='answer-alert'>{isCorrectAnswer()}</p> : null}

                </div> : <div className='results-container'>
                    <div className='results-box'>
                        <p className='results-score'>Congratulations! You scored {score} / {MAX_QUESTIONS}</p>
                        <p className='play-again'>Play Again?</p>
                        <button className='n-btn n-btn-styling' onClick={() => resetState()}>Yes</button>
                        <button className='n-btn n-btn-styling' onClick={() => setGameStatus(false)}>No</button>
                    </div>
                </div>}
        </div>
    )

}

export default TriviaCard