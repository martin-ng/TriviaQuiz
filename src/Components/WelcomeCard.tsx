import React, { FC } from 'react'
import '../css/style.css'

type Props = {
    setGameStatus: any;
}

const WelcomeCard: FC<Props> = ({ setGameStatus }) => {

    return (
        <div className="welcome-container">
            <h3 id='welcome'>Welcome to the Quiz App</h3>
            <button className="wc-btn-styling" onClick={() => setGameStatus(true)}>Play</button>
        </div >
    )
}

export default WelcomeCard