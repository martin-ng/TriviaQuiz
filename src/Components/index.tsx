// This file acts as a central directory for all components

export { default as TriviaCard } from './TriviaCard'
export { default as WelcomeCard } from './WelcomeCard'
export { default as AnswerCard } from './AnswerCard'