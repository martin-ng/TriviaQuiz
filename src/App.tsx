import React, { useState } from 'react';
import "./css/App.styles.css"

// Components
import { WelcomeCard, TriviaCard } from './Components/index'

const App = () => {
  const [gameStatus, setGameStatus] = useState<boolean>(false)

  return (
    <div>
      {gameStatus ? <TriviaCard setGameStatus={setGameStatus} /> : <WelcomeCard setGameStatus={setGameStatus} />}
    </div>

  );
}

export default App;
