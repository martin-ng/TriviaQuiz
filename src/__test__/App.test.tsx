import React from 'react';
import { screen } from "@testing-library/react";
import { WelcomeCard } from '../Components/'
import App from '../App';

import { render } from '@testing-library/react';
import { shallow } from 'enzyme';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });


describe("<App /> component", () => {

    it("renders correctly without crashing", () => {
        const wrapper = shallow(<WelcomeCard />);
        expect(wrapper).toMatchSnapshot()
    })

    it("displays `Welcome to the Quiz App`", () => {
        render(<App />)
        const textElement = screen.getByText(/Welcome to the Quiz App/i);
        expect(textElement).toBeInTheDocument();
    })
})

