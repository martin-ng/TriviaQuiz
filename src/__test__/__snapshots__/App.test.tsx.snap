// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> component renders correctly without crashing 1`] = `ShallowWrapper {}`;

exports[`<App /> component renders welcome without crashing 1`] = `ShallowWrapper {}`;
